### ISSUE TITLE
Version 3.0 leaks threads

### ISSUE THREAD

**Issue by phoerious** [hangoutsbot/hangoutsbot#866]
## Description
I recently upgraded to the 3.0 beta version of this bot. Shortly thereafter I had to realize that cron jobs started failing, because they ran out of file handles, websites went down intermittently and I couldn't even start a new shell anymore.

Basically, everything that spawned a process failed with an error like

-bash: fork: retry: Resource temporarily unavailable

A look at the process list showed a shit load of bot threads running. Each bot still only had one main process, but a gazillion running threads.

Version 2.8 did not have this problem.

## Output of `/bot version`

Bot Version: 3.0.0-beta
Python Version: 3.6.1
* aiohttp 2.2.3
* appdirs 1.4.3
* emoji 0.4.5
* hangups 0.4.2
* telepot 12.1

I'm running the bot via daemontools and the startup script looks like this:

```
exec $HOME/.local/bin/hangupsbot --config $HOME/etc/hangoutsbot/bot/config.json --memory $HOME/etc/hangoutsbot/bot/memory.json --cookies $HOME/etc/hangoutsbot/bot/cookies.json --log /dev/null 2>&1
```

**Comment by das7pad** [comment]
@phoerious does this issue also exist when running the bot in a terminal?

Can you provide a terminal and log output of a complete run on level info?
This should include the bot startup, ``/bot resourcememory`` in at least two different chats, stop via ``KeyboardInterrupt`` or ``/bot quit`` and the unload of the bot. Please attach the files then to keep the issue clean.

**Comment by phoerious** [comment]
Logs are here:

INFO: https://pastebin.com/vDndsrYp
STDOUT / STDERR: https://pastebin.com/XhV1hk3j

They are almost identical. The huge stacktrace tail came after I quit with Ctrl+C.

I suppose most interesting is this message:

@400000005960b01113be310c 12:12:23 ERROR asyncio: Unclosed client session
@400000005960b01113be4494 client_session: <aiohttp.client.ClientSession object at 0x7f5ac8b75e48>

The daemontools STDOUT logs are also littered with this message. It seems to repeat again and again the longer the bot is running.

**Comment by das7pad** [comment]
The ``aiohttp.ClientSession`` messages are introduced in aiohttp version 2.2, running ``$ /path/to/venv/bin/pip3 install  --upgrade aiohttp==2.1`` should fix this.

Can you please run the bot again and use ``... --log log > stdout`` instead of ``tee``? the pipe to ``tee`` is closed before the bot finished unloading.

The diff of your ``/bot resourcememory`` output shows a huge memory leak, I also saw when using the ``v3.0`` branch.

**Comment by phoerious** [comment]
I updated the pastes.

**Comment by phoerious** [comment]
So, an hour later, one of my bots already has 20 threads, another has 14, the other less active bots have between 5 and 8 each. The machine has 12 hyper threads, so apparently the number of threads is not limited by the number of cores. 122 threads isn't dangerous, but compared to what the bot is supposed to do quite unnecessary. My bots are busy, but not THAT busy. I don't know exactly how the core routines work, but 3-5 constantly running threads per bot should be enough (especially since asyncio is single-threaded anyway).

**Comment by phoerious** [comment]
Alright. After two of my bots passed the 60 threads mark, I set up an hourly cron job to restart them periodically until this is resolved.

**Comment by das7pad** [comment]
20 threads are peanuts for your system and so far the maximum I saw when using ``asyncio`` and ``aiohttp`` in concurrent calls. Running the [1mio requests with aiohttp](http://www.artificialworlds.net/blog/2017/06/12/making-100-million-requests-with-python-aiohttp/) script spawns only 20 threads to file the concurrent requests.

I thought we are talking in greater numbers, you can close this issue.

_From my observations: Once the threads are spawned by ``asyncio`` they are not closed before the termination of the main script._
Having 20 threads proves that async calls are usefull as 20 concurrent http request were made at least once.

The command ``ulimit -u <int>`` can be used to increase the number of allowed [processes -> threads](https://stackoverflow.com/questions/344203/maximum-number-of-threads-per-process-in-linux/344292#344292) on your host, or you can edit the target manually: ``/etc/security/limits.d/20-nproc.conf``.

You can use ``htop`` and the tree-view (press ``T`` or ``F5``) to display the activity of the spawned threads, only a few of them will show continuous activity.

**Comment by das7pad** [comment]
I missed the last comment somehow.

Does this issue also exist when using python version 3.5.x?

I am running a custom version based on v3.0 and python 3.5.3. Having a full sync stack with ``HO<->HO<->TG<->Slack`` does not _leak_ threads further the mentioned 20 threads.

**Comment by phoerious** [comment]
As I said, 20 threads aren't critical, but unnecessary. For a classic pre-forking web server 20 constantly running threads would be normal, but the bot is not a classic web server and a few milliseconds extra latency for spawning a new thread on demand really don't make a difference. Furthermore, it doesn't behave like a pre-forking web server. Such a server would

* spawn a constant thread pool at startup
* re-use threads where possible
* only spawn excess threads when needed
* destroy any excess threads some time after they are done
* have an upper limit of how many threads can be spawned
* periodically replace running threads with new ones to avoid resource leakage

The bot doesn't exhibit any of those properties:

* it does not pre-fork (starts with only 3-4 threads)
* it apparently doesn't re-use any threads (otherwise it wouldn't constantly need new threads)
* it doesn't destroy any threads (the number only goes up, never down)
* it doesn't have any conceivable thread limit (even for the most active bot, 60+ long-lived threads are just ridiculous)

Also those threads would ideally run concurrently on different CPU cores, which in Python they don't. So even if you had 300 threads running, still only one request could be served at a time (although with flexible CPU scheduling), thanks to Python being an interpreted language.

One of my most active bots (which is also one of those bots with the highest thread count), isn't actually in very many hangouts. It only syncs three hangouts with three telegrams. So even during the most heated conversations, it would never need more than three threads (+ maybe a couple for fetching images from links etc.). However, it keeps spawning new threads without closing any old ones.

Long story short...
I restarted one of the bots with Python 3.5. Will report back in a couple of hours.

**Comment by phoerious** [comment]
It's been three hours now. One bot has 11 threads, another 12, one 22. Rest around 5-6. The thread count therefore seems to rise much slower and has possibly settled at this number. So it seems that the thread leak problem mainly appears with Python 3.6. However, the fact that the thread count, although, much slower, still only increases and never decreases, still makes me think that the bot is never actually closing any threads.

**Comment by das7pad** [comment]
I am not aware of a funtion to control the threads spawned by ``asyncio`` from inside a single event loop.

I wrote a [tester](https://gist.github.com/das7pad/f2ae50e60a5ca2d6b8a371645deb5549) which fetches async resources split in multiple queues (can be seen as multiple sources for messages). The [first version](https://gist.github.com/das7pad/f2ae50e60a5ca2d6b8a371645deb5549/7a507b6956a0b288574a80f6cce92cb8c6e02783) runs a single loop (similar to the behaviour of the bot) and more threads are added on each run. The [second version](https://gist.github.com/das7pad/f2ae50e60a5ca2d6b8a371645deb5549/7dd386bdff57dca485472fd5170fc19c37033893) closes the loop after each run (this is not possible with the bot during runtime) and discards the previous threads.



How many threads are spawned on your host when using more than 20 (and below 50, or adjust the ``TOTAL_REQUESTS``) queues?

**Comment by phoerious** [comment]
The first script spawns one thread per request (so with 5*10 I end up with 51 threads in the end). However, it seems to be capped at 60 worker threads (i.e. 61 threads total). After that, no more threads are spawned. Tested it with both Python 3.5 and 3.6.

The second script doesn't work. You close the event loop after the first iteration, so nothing is executed anymore in the following runs.

**Comment by das7pad** [comment]
Oh, I forgot to update the part before the try block.
https://gist.github.com/das7pad/f2ae50e60a5ca2d6b8a371645deb5549/7abd567253eb8a0178bcdef5d8a43f2661a400f5

**Comment by phoerious** [comment]
Yeah, that way the threads are destroyed after every iterations.

**Comment by phoerious** [comment]
It seems clear to me now, that no matter what, every bot will at some point reach that 61 threads mark. One of my bots just did (/bot resourcememory gives memory (resource): 63.83984375 MB btw).
So with my 5 bots running, that is 305 threads after a day or two. It isn't super critical, but definitely a bug that must be fixed.

On the bright side, I haven't yet had any problems with running out of file handles again. So I suppose that was a combination of the thread issue and the aiohttp 2.2 problem where each of those 60 threads also leaked file handles for unclosed HTTP connections.

**Comment by das7pad** [comment]
I searched a few hours in stdlib/``asyncio`` code and in asyncio related issues/articles, but found nothing useful for this issue.
From the test results I would rather blame ``asyncio`` instead of this project for the persistant (inactive) threads.

You could file a bug report at python or create a question on Stack Overflow to limit the number of spawned threads in a specific python process.

I do not think that we should discard the advantage of detached event/message handling just to limit the number of spawned threads.

**Comment by phoerious** [comment]
I stand corrected: those 305 threads are enough for creating the fork issue. This is really a problem.

**Comment by das7pad** [comment]
@phoerious Did you increased the process limit for the user which is used to run the bots?

**Comment by phoerious** [comment]
I can't. This is a multi-user system for which I don't have root permissions.

**Comment by phoerious** [comment]
I found something. Apparently the thread limit is platform-specific. On my home machine, the limit is 40 workers, not 60. On your system it's probably even lower, since you have never experienced any problems and your thread count stays at about 20.

I tried changing the default executor like this:

```
loop.set_default_executor(ThreadPoolExecutor(max_workers=5))
```
and it actually worked. Now I have a maximum of 6 threads.

**Comment by phoerious** [comment]
I created a pull request. Please have a look: #866

**Comment by das7pad** [comment]
Looking at your fix in detail, I found [this](https://github.com/python/cpython/blob/master/Lib/asyncio/base_events.py#L631..L632) and [this](https://github.com/python/cpython/blob/master/Lib/concurrent/futures/thread.py#L100) part, which set the thread limit if none was specified:
```
@asyncio.BaseEventLoop.run_in_executor(self, executor, ...)
    if executor is None:
        executor = self._default_executor
        if executor is None:
            executor = concurrent.futures.ThreadPoolExecutor()
            self._default_executor = executor
```
and
```
@concurrent.futures.ThreadPoolExecutor.__init__(self, max_workers=None, ...)
    if max_workers is None:
        max_workers = (os.cpu_count() or 1) * 5
```
This explains the different thread numbers:
- The host you run the bots on: 12 CPUs <-> 60Threads
- Your home maschine 40Threads -> 8CPUs
- My host having 4CPUs <-> 20Threads

[Here](https://github.com/python/cpython/blob/master/Lib/concurrent/futures/thread.py#L130..L133) is a TODO note to avoid creating new threads:
```
@concurrent.futures.ThreadPoolExecutor._adjust_thread_count(self)
    # TODO(bquinlan): Should avoid creating new threads if there are more
    # idle threads than items in the work queue.
    num_threads = len(self._threads)
    if num_threads < self._max_workers:
        ...create new thread
```

**Comment by phoerious** [comment]
Nice find. Good thing I'm not running the bots on our 32 core University cluster machines. :-D

**Comment by vaindil** [comment]
I'm on 3.0 and am not having this problem, happy to provide any info that could help.

> BTW I don't remember how much memory the old version consumed, but my bots are in quite a few hangouts. 45MB wouldn't surprise me.

I've never known how much memory mine "should" be using... it's in 194 groups and uses 276MB.

**Comment by das7pad** [comment]
@vaindil your comment disappeared somehow.
> it's in 194 groups and uses 276MB

That's quite high, which plugins are you using? uptime?

_I am running a custom version: One of my bots is member in 434 conversations (with 679 seen Users), running telesync and has a memory footprint of 124MB after about one week uptime and ~900 handled messages._

**Comment by vaindil** [comment]
@das7pad I deleted it almost immediately, I realized it was completely off-topic for this issue. I only posted it because I saw that comment mentioning it above.

My uptime at the moment is about 10 days. I definitely have more than 900 messages in that time, but I don't know how to check.

**Comment by das7pad** [comment]
@vaindil Agree, we should better open a new issue for the memory leak.
